privacyIDEAADFSProvider/Adapter.cs

```diff
@@ -210,64 +220,71 @@ public IAuthenticationAdapterMetadata Metadata
             }
 
             // Do the authentication according to the mode we are in
-            if (privacyIDEA != null)
+            PIResponse response = null;
+            if (mode == "push")
             {
-                PIResponse response = null;
-                if (mode == "push")
+                if (privacyIDEA.PollTransaction(transactionid))
                 {
-                    if (privacyIDEA.PollTransaction(transactionid))
-                    {
-                        // Push confirmed, finish the authentication via /validate/check using an empty otp
-                        // https://privacyidea.readthedocs.io/en/latest/tokens/authentication_modes.html#outofband-mode
-                        response = privacyIDEA.ValidateCheck(user, "", transactionid);
-                        if (response != null && response.Value)
-                        {
-                            outgoingClaims = Claims();
-                            return null;
-                        }
-                    }
-                    else
-                    {
-                        // Else push not confirmed yet
-                        form.ErrorMessage = "Authenication not confirmed yet!";
-                    }
+                    // Push confirmed, finish the authentication via /validate/check using an empty otp
+                    // https://privacyidea.readthedocs.io/en/latest/tokens/authentication_modes.html#outofband-mode
+                    response = privacyIDEA.ValidateCheck(user, "", transactionid);
                 }
-                else if (mode == "webauthn")
+                else
                 {
-                    // TODO webauthn implementaiton
+                    // Else push not confirmed yet
+                    form.ErrorMessage = "Authenication not confirmed yet!";
+                }
+            }
+            else if (mode == "webauthn")
+            {
+                string origin = (string)GetFromDict(proofDict, "origin");
+                string webauthnresponse = (string)GetFromDict(proofDict, "webAuthnSignResponse");
+
+                if (string.IsNullOrEmpty(origin) || string.IsNullOrEmpty(webauthnresponse))
+                {
+                    Error("Incomplete data for WebAuthn authentication: WebAuthnSignResponse or Origin is missing!");
+                    form.ErrorMessage = "Could not complete WebAuthn authentication. Try again or use another token type.";
                 }
                 else
                 {
-                    // Mode == OTP
-                    response = privacyIDEA.ValidateCheck(user, otp, transactionid);
-                    if (response != null && response.Value)
-                    {
-                        outgoingClaims = Claims();
-                        return null;
-                    }
+                    response = privacyIDEA.ValidateCheckWebAuthn(user, transactionid, webauthnresponse, origin);
                 }
+            }
+            else
```

Comment:
I would go for 

    else if (mode == "interactive")

and then finally do

    else {
       raise Exception("Mode not supported").
    }

This might be more clear, and then is allowed to fail with a u2f token.

Reply:
This is not the mode that will be sent from privacyIDEA in the future. 
It is the internal mode of the plugin, therefore there can be no mode that is not supported.

Reply:
Ah, ok. YOu are parsing the mode somewhere else. So at this point the plugin already knows the mode.

Conceptually I decided in the JSON Response to all the default mode "interactive" and not "otp", since it can be anything - like entering a PIN to be changed.

I would appreciate, if you also would call it here in the code "interactive" and not "otp", to have a uniform wording.
(The Glossry is to be done :-)